packages/drivers/odsp-driver/src/epochTracker.ts

```diff
@@ -0,0 +1,156 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import assert from "assert";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { createOdspNetworkError, fluidEpochMismatchError, OdspErrorType } from "@fluidframework/odsp-doclib-utils";
+import { fetchAndParseAsJSONHelper, fetchHelper, IOdspResponse } from "./odspUtils";
+import { ICacheEntry, LocalPersistentCacheAdapter } from "./odspCache";
+
+/**
+ * This class is a wrapper around fetch calls. It adds epoch to the request made so that the
+ * server can match it with its epoch value in order to match the version.
+ * It also validates the epoch value received in response of fetch calls. If the epoch does not match,
+ * then it also clears all the cached entries for the given container.
+ */
+export class EpochTracker {
+    private _fluidEpoch: string | undefined;
+    private _hashedDocumentId: string | undefined;
+    constructor(
+        private readonly persistedCache: LocalPersistentCacheAdapter,
+        private readonly logger?: ITelemetryLogger) {
+    }
+
+    public set hashedDocumentId(docId: string | undefined) {
+        assert(this._hashedDocumentId === undefined, "DocId should be set only once");
+        assert(docId !== undefined, "Passed docId should not be undefined");
+        this._hashedDocumentId = docId;
+    }
+
+    public get hashedDocumentId(): string | undefined {
+        return this._hashedDocumentId;
+    }
+
+    public get fluidEpoch() {
+        return this._fluidEpoch;
+    }
+
+    public async fetchFromCache<T>(entry: ICacheEntry, maxOpCount: number | undefined): Promise<T | undefined> {
+        const value = await this.persistedCache.get(entry, maxOpCount);
+        if (value !== undefined) {
+            try {
+                this.validateEpochFromResponse(value.fluidEpoch);
```

Comment:
If that fails epoch test, we will clear cache, but we will still return caches value on line 48. This does not look right.

Reply:
Yeah needed to throw the error.